### PR TITLE
ng_at86rf2xx: Implement CSMA en/disable and setting retries

### PR DESCRIPTION
**Pull request by daniel-k**
When enabling CSMA it defaults to 4 retries and sets min BE to 3 and max BE 5. I'm not sure if these are sensible defaults, they are the reset values of the transceiver. 

The BE cannot yet be configured though. I wasn't sure how to integrate the BE setting into the given interfaces, so any proposal is welcome!

I'm not sure if there is a way to test and verify this in the current state, but I've got more changes in the pipe that depend on this (namely TX feedback, see https://github.com/RIOT-OS/RIOT/issues/3125).

### CODE REVIEW

File: drivers/kw2xrf/kw2xrf.c [hunk 1]
```diff
@@ -754,7 +754,7 @@ int kw2xrf_set(ng_netdev_t *netdev, ng_netconf_opt_t opt, void *value, size_t va
                               ((bool *)value)[0]);
             return sizeof(ng_netconf_enable_t);
 
-        case NETCONF_OPT_AUTOCCA:
+        case NETCONF_OPT_CSMA:
```
Comment (OlegHahm) [inline drivers/kw2xrf/kw2xrf.c]: Is this change intensional?

**Comment by jremmert-phytec-iot** [comment]
> When enabling CSMA it defaults to 4 retries and sets min BE to 3 and max BE 5. I'm not sure if these are sensible defaults, they are the reset values of the transceiver.

The IEEE 802.15.4 (2011) standard also defines them as standard values. However, if one want to send out data at a precise time, without (unpredictable) delay, BE should be set to 0. Figure 11 (p. 23) in the IEEE 802.15.4 standard shows that the backoff-wait is performed always initially before CCA (and of course later again, if the channel was busy).

Since this behaviour could be of interest for other MAC-protocols and is hidden "hidden" in HW in this Atmel device, we should maybe document that somewhere in the driver. I think e.g. Contiki-MAC would expect a min BE value of 0 to avoid any delays before sending.
 
The wait-time, after which the packet will be send again, is defined as:
 * ```random(2^(BE–1)) * backoffPeriod```
 *  ```backoffPeriod = 320µs```

**Comment by daniel-k** [comment]
> Since this behaviour could be of interest for other MAC-protocols and is hidden "hidden" in HW in this Atmel device, we should maybe document that somewhere in the driver. I think e.g. Contiki-MAC would expect a min BE value of 0 to avoid any delays before sending.

That's actually a very good point. I wasn't aware that the random backoff wait is already active before the first attempt to transmitting. As I'm aiming for a MAC layer is even more important to me.

So in the first step it would be ok for **me** to have minBE=0 as a default, but I guess this should be configurable somehow. While going through all the interfaces involved, I still fail to find a good place for this, as it's really hardware specific and maybe not well placed as a NETCONF option.

What do you say?

**Comment by jremmert-phytec-iot** [comment]
yes, I guess making this option configurable is a good idea. 

I actually think it should be an ordinary NETCONF option. All devices that implements CSMA/backoff in hardware will be able to set this option. All other devices simply don't implement it and will automatically return -ENOTSUP by design. But we should clarify in the documentation, that this option is explicitly described as hardware-only function. If the option can not be set, the MAC layer has to handle the CSMA/backoff procedure. @haukepetersen do you agree?

To go further, this could even be an indication for the MAC layer to en/disable software CSMA/backoff.

**Comment by OlegHahm** [comment]
I'm without a computer for the next two and a half weeks - hence, someone else should review this.

**Comment by daniel-k** [comment]
@OlegHahm ping

File: drivers/include/ng_at86rf2xx.h [hunk 2]
```diff
@@ -338,6 +338,48 @@ uint8_t ng_at86rf2xx_get_max_retries(ng_at86rf2xx_t *dev);
...
+ *
+ * @return                  configured number of retries
+ */
+uint8_t ng_at86rf2xx_get_csma_max_retries(ng_at86rf2xx_t *dev);
+
+/**
+ * @brief   Set the maximum number of channel access attempts per frame (CSMA)
+ *
+ * This setting specifies the number of attempts to access the channel to
+ * transmit a frame. If the channel is busy *retries* times, then frame
```
Comment (OlegHahm) [inline drivers/include/ng_at86rf2xx.h]: `@p retries`

File: sys/include/net/ng_netconf.h [hunk 3]
```diff
@@ -119,8 +119,12 @@ typedef enum {
      * @note not all transceivers may support this interrupt
      */
     NETCONF_OPT_TX_END_IRQ,
-    NETCONF_OPT_AUTOCCA,            /**< en/disable to check automatically
-                                         before sending the channel is clear. */
+    NETCONF_OPT_CSMA,               /**< en/disable to check automatically
```
Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: Why did you rename the option?

File: drivers/ng_at86rf2xx/ng_at86rf2xx_getset.c [hunk 4]
```diff
@@ -249,6 +249,56 @@ void ng_at86rf2xx_set_max_retries(ng_at86rf2xx_t *dev, uint8_t max)
...
+    tmp  = ng_at86rf2xx_reg_read(dev, NG_AT86RF2XX_REG__XAH_CTRL_0);
+    tmp &= NG_AT86RF2XX_XAH_CTRL_0__MAX_CSMA_RETRIES;
+    tmp >>= 1;
+    return tmp;
+}
+
+void ng_at86rf2xx_set_csma_max_retries(ng_at86rf2xx_t *dev, int8_t retries)
+{
+    retries = (retries > 5) ? 5 : retries; // valid values: 0-5
+    retries = (retries < 0) ? 7 : retries; // max < 0 => disable CSMA (set to 7)
```
Comment (OlegHahm) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_getset.c]: No C++ comments, please.

File: drivers/ng_at86rf2xx/ng_at86rf2xx_getset.c [hunk 5]
```diff
@@ -249,6 +249,56 @@ void ng_at86rf2xx_set_max_retries(ng_at86rf2xx_t *dev, uint8_t max)
...
+void ng_at86rf2xx_set_csma_seed(ng_at86rf2xx_t *dev, uint8_t entropy[2])
+{
+    if(entropy == NULL) {
+        DEBUG("[ng_at86rf2xx] opt: CSMA seed entropy is nullpointer");
+        return;
+    }
+    DEBUG("[ng_at86rf2xx] opt: Set CSMA seed to 0x%x 0x%x", entropy[0], entropy[1]);
+
+    ng_at86rf2xx_reg_write(dev,
+        NG_AT86RF2XX_REG__CSMA_SEED_0,
```
Comment (OlegHahm) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_getset.c]: indentation is off

File: sys/include/net/ng_netconf.h (hunk 3 above)
Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: I also added `NETCONF_OPT_CSMA_RETRIES` and to make clear that this is linked to `NETCONF_OPT_AUTOCCA` I figured that it should be renamed. Only the kw2xrf is using this option and this board also does CSMA. It's more precise and googling `AUTOCCA` only yields rubbish. 
But I see that renaming legacy options might not be a good idea per se.

Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: Well, I actually thought that `AUTOCCA` is more to the point, since `OPT_CSMA` somehow suggest that this completely disable and enables the MAC protocol.

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: When you say "MAC protocol", what do you mean exactly by this? CSMA/CA is exactly the process of sensing the channel and depending on the outcome (and maybe a random backoff) the decision to send or not.

As far as I understand what you mean by "AUTOCCA", this seems equal.

Comment (jremmert-phytec-iot) [inline sys/include/net/ng_netconf.h]: We should distinguish between the terms CCA and CSMA. CCA only senses the channel for an ongoing transmission.  CSMA/CA, however, is the complete process that includes backoff handling. CCA is rather a part of the CSMA/CA process.

The differences between this device and the kw2x device is that the kw2x device does not integrate the mechanisms for backoff handling in hardware. Though it integrates an automatic CCA check before sending out a packet. Thats the reason why I had chosen the name AUTOCCA.

What do you think of renaming NETCONF_OPT_AUTOCCA to NETCONF_OPT_CCA and introduce NETCONF_OPT_CSMA as a new option. The kw2x device e.g. does not implement the option NETCONF_OPT_CSMA. This would allow to en/disable SW-CSMA in the MAC-layer according to the devices' capabilities.

Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: > We should distinguish between the terms CCA and CSMA. CCA only senses the channel for an ongoing transmission. CSMA/CA, however, is the complete process that includes backoff handling. CCA is rather a part of the CSMA/CA process.

That's exactly what I've meant. Thanks for the elaboration, @jremmert-phytec-iot. I was a little too tired yesterday to express myself clearer.

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: > What do you think of renaming NETCONF_OPT_AUTOCCA to NETCONF_OPT_CCA and introduce NETCONF_OPT_CSMA as a new option. The kw2x device e.g. does not implement the option NETCONF_OPT_CSMA. This would allow to en/disable SW-CSMA in the MAC-layer according to the devices' capabilities.

I like that. But how to ensure that these options don't interfere? I could imagine to implement AUTOCCA for the at86rf2xx by setting minBE = maxBE = 0 and CSMA_RETRIES = 0. But then you shouldn't mess with the CSMA setting.

When implementing SW-CSMA, one would try enabling CSMA first, and if that fails, try AUTOCCA I guess.

What is the bahaviour of kw2xrf? Is it trying resend the packet when channel is not clear? Anyway we need signaling if the packet was actually sent. I have an implementation for that, but didn't PR yet.

Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: @jremmert-phytec-iot
> The differences between this device and the kw2x device is that the kw2x device does not integrate the mechanisms for backoff handling in hardware. Though it integrates an automatic CCA check before sending out a packet.

I don't quite get this part of your comment. What exactly does the kw2x when its integrated CCA detects that the channel is already occupied?

Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: @daniel-k 
> I could imagine to implement AUTOCCA for the at86rf2xx by setting minBE = maxBE = 0 and CSMA_RETRIES = 0. 

Why would you set `minBE = maxBE = 0` for CCA? In my understanding just checks if the medium is currently free for sending. If this is not the case, the transceiver initiates another channel sensing according to the configured backoff time. So why would you fix this backoff time if CCA is enabled and why would you disable retransmissions in this case?

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: > Why would you set minBE = maxBE = 0 for CCA? In my understanding just checks if the medium is currently free for sending. If this is not the case, the transceiver initiates another channel sensing according to the configured backoff time. So why would you fix this backoff time if CCA is enabled and why would you disable retransmissions in this case?

To emulate what kw2xrf is doing. `minBE = maxBE = 0` effectively turns off (random) backoff, so that the frame is immediately sent if the channel is clear. Turning of retries is what I speculated is the behaviour of kw2xrf, because immediately sensing again makes no sense.

Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: > so that the frame is immediately sent if the channel is clear

Doesn't this value back off the time for sensing the channel? If the channel is sensed clear, the packet should always be sent immediately, right?

But otherwise, after looking at the datasheet again, it makes sense, what you've written. I confused CSMA_RETRIES with FRAME_RETRIES.

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: > Doesn't this value back off the time for sensing the channel? If the channel is sensed clear, the packet should always be sent immediately, right?

Nope ;) Have a look at the 802.15.4 specification. See page 172 of the 2006 version (Figure 69) for the actual CSMA/CA algorithm. Before every channel sensing, there's the random backoff waiting.

Comment (OlegHahm) [inline sys/include/net/ng_netconf.h]: I don't have the 2006 version, but to my understanding that's exactly what I thought: sensing the channel is backed off, not the sending.

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: Okay, you're right. Instead of
> so that the frame is immediately sent if the channel is clear. 

it should be

> so that the channel is sensed immediately without random backoff.

Comment (jremmert-phytec-iot) [inline sys/include/net/ng_netconf.h]: > But how to ensure that these options don't interfere? I could imagine to implement AUTOCCA for the at86rf2xx by setting minBE = maxBE = 0 and CSMA_RETRIES = 0. But then you shouldn't mess with the CSMA setting.

True, that could be ambiguous. Then maybe we could just introduce the option NETCONF_OPT_CCA for CCA. CSMA could be de/activated by setting the options minBE, maxBE and CSMA_RETRIES. Do you see any option clash there?

> When implementing SW-CSMA, one would try enabling CSMA first, and if that fails, try AUTOCCA I guess.

Yes, thats what I also had in mind.

> What is the bahaviour of kw2xrf? Is it trying resend the packet when channel is not clear? Anyway we need signaling if the packet was actually sent. I have an implementation for that, but didn't PR yet.

> What exactly does the kw2x when its integrated CCA detects that the channel is already occupied?

@OlegHahm @daniel-k The kw2x does not try to resend the packet in that case. It just triggers an interrupt; with a read of the status register one can figure out the ISR-reason then.

In my currently open CSMA_PR (#2467) this is handled by introducing the new event-types NETDEV_EVENT_CCA_CHANNEL_BUSY/IDLE. The driver has to figure out according to the occured interrupts what event it needs to trigger.

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: > True, that could be ambiguous. Then maybe we could just introduce the option NETCONF_OPT_CCA for CCA. CSMA could be de/activated by setting the options minBE, maxBE and CSMA_RETRIES. Do you see any option clash there?

No option clash, but CSMA should be explicitly be enabled or disabled, not via CSMA_RETRIES. And what would `NETCONF_OPT_CCA` actually do? In this case `AUTOCCA` is more understandable.

> In my currently open CSMA_PR (#2467) this is handled by introducing the new event-types NETDEV_EVENT_CCA_CHANNEL_BUSY/IDLE. The driver has to figure out according to the occured interrupts what event it needs to trigger.

There is already `NETCONF_OPT_IS_CHANNEL_CLR` to manually sense the channel. Otherwise I find it more logical to give feedback like `NETDEV_EVENT_TX_MEDIUM_BUSY` when sending was not possible (i.e. channel not clear) or just `NETDEV_EVENT_TX_COMPLETE` if the frame was sent out. (that's how I've done [it](https://github.com/daniel-k/RIOT/blob/4cf2f53010a975e650f4093b757ef2c043c053d2/sys/net/link_layer/ng_lwmac/ng_lwmac.c#L392))

File: sys/include/net/ng_netconf.h [hunk 6]
```diff
@@ -119,8 +119,12 @@ typedef enum {
...
+                                     *   before sending the channel is clear. */
+    NETCONF_OPT_CSMA_RETRIES        /**< get/set the number of retries when
```
Comment (thomaseichinger) [inline sys/include/net/ng_netconf.h]: Please add a semicolon, there were other options added below in the meanwhile.

File: drivers/ng_at86rf2xx/ng_at86rf2xx_getset.c [hunk 7]
```diff
@@ -297,8 +351,9 @@ void ng_at86rf2xx_set_option(ng_at86rf2xx_t *dev, uint16_t option, bool state)
         /* trigger option specific actions */
         switch (option) {
             case NG_AT86RF2XX_OPT_CSMA:
-                DEBUG("[ng_at86rf2xx] opt: disabling CSMA mode (NOT IMPLEMENTED)\n");
-                /* TODO: en/disable csma */
+                DEBUG("[ng_at86rf2xx] opt: disabling CSMA mode\n");
+                // Setting retries to -1 means CSMA disabled
```
Comment (thomaseichinger) [inline drivers/ng_at86rf2xx/ng_at86rf2xx_getset.c]: Please change comment style.

File: sys/include/net/ng_netconf.h (hunk 3 above)
Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: @jfischer-phytec-iot @OlegHahm 

How to proceed?

Comment (jremmert-phytec-iot) [inline sys/include/net/ng_netconf.h]: > No option clash, but CSMA should be explicitly be enabled or disabled, not via CSMA_RETRIES. And what would NETCONF_OPT_CCA actually do? In this case AUTOCCA is more understandable.

Yes I agree, makes it more clear. Just thought you'd prefer the name NETCONF_OPT_CCA instead of AUTOCCA.

I don't really know how to proceed. (i) Using both options ..CCA and ..CSMA does not really fit to a practical implementation related to the atmel device, right? And (ii), Omitting the ..CCA option would fit to an implementation on the atmel device, but that would be very ambiguous when it comes to the KW2x device since it actually does not support CSMA.

What I need for KW2x device is an option that allows enabling the CCA part. The CSMA related options can not be implemented on the KW2x device anyways.

Comment (daniel-k) [inline sys/include/net/ng_netconf.h]: > Using both options ..CCA and ..CSMA does not really fit to a practical implementation related to the atmel device, right?

Depends. As I said, the `AUTOCCA` behaviour can be emulated (in HW). And as a SW-CSMA would test for `NETCONF_OPT_CSMA` first and then falling back to `NETCONF_OPT_AUTOCCA` there would be no problem either. But we must ensure that the documentation really makes clear the difference.

> Omitting the ..CCA option would fit to an implementation on the atmel device, but that would be very ambiguous when it comes to the KW2x device since it actually does not support CSMA.

Another argument for having both coexist for the moment.

Comment (jremmert-phytec-iot) [inline sys/include/net/ng_netconf.h]: > Depends. As I said, the AUTOCCA behaviour can be emulated (in HW). And as a SW-CSMA would test for NETCONF_OPT_CSMA first and then falling back to NETCONF_OPT_AUTOCCA there would be no problem either. But we must ensure that the documentation really makes clear the difference.

Ok great, then I would vote for that solution.

**Comment by daniel-k** [comment]
@jremmert-phytec-iot I updated the docs and now _AUTOCCA and _CSMA options exist both in parallel. But there is no "AUTOCCA-emulation" for at86rf2xx yet.

**Comment by thomaseichinger** [comment]
Please rebase.

**Comment by OlegHahm** [comment]
@daniel-k, willing to provide an extension to the `ifconfig` shell command to control this? If not, please rebase, squash, and send me a ping so I can test it.

**Comment by daniel-k** [comment]
Rebased. Will look into `ifconfig` now.

**Comment by daniel-k** [comment]
@OlegHahm 
`sc_netif.c` has functions for setting `u16` und `i16` sized netopt options, while my new option `NETOPT_CSMA_RETRIES` expects `uint8_t`. I wonder now if I should expand my option to `uint16_t` or if I should add another function to `sc_netif.c` handling `uint8_t`. That would mostly depend on what those `NETOPT` options should look like in general.

**Comment by OlegHahm** [comment]
I would suggest to add either a setter for `uint8_t` to the shell commands or see if you can come up with a unified setter function that take different widths.

**Comment by daniel-k** [comment]
@OlegHahm 
Shell controls are now implemented for enabling/disabling CSMA and setting retries. Waiting for (N)ACK now.

**Comment by OlegHahm** [comment]
ACK. Please rebase and squash!

**Comment by OlegHahm** [comment]
You have to add a string representation of the new netopt types.

**Comment by daniel-k** [comment]
@OlegHahm string representations added and squashed again
